docs/book/start/parsing-packets.md

```diff
@@ -0,0 +1,100 @@
+# Parsing Packets
+
+In the previous chapter, our XDP application ran until Ctrl-C was hit and permitted
+all the traffic. Each time a packet was received, the BPF program created a logged
+the string "received a packet" for each packet received. In this chapter we're
+going to show how to parse packets.
+
+While we could go all out and extract data all the way up to L7, we'll constrain
+our example to L3, and to make things easier, IPv4 only.
+
+!!! example "Source Code"
+
+    Full code for the example in this chapter is available [here](https://github.com/aya-rs/book/tree/main/examples/xdp-log)
+
+## Using Network Types
+
+To get useful data to log, we first need some useful data structures to
+to populate with data from the `XdpContext`.
+We want to log the Source IP Address of incoming traffic, so we'll need to:
+
+1. Read the Ethernet Header to determine if this is an IPv4 Packet
+1. Read the Source IP Address from the IPv4 Header
+
+The [network-types crate](https://crates.io/crates/network-types) provides
+networking structs, including `EthHdr` and `Ipv4Hdr` which we are ging to use
+in our program.
+
+Let's add it to our eBPF crate by adding a dependency on `network-types` in our
+`xdp-log-ebpf/Cargo.toml`:
+
+=== "xdp-log-ebpf/Cargo.toml"
+
+    ```toml linenums="1"
+    --8<-- "examples/xdp-log/xdp-log-ebpf/Cargo.toml"
+    ```
+
+## Getting Packet Data From The Context And Into the Map
+
+The `XdpContext` contains two fields, `data` and `data_end`.
+`data` is a pointer to the start of the data in kernel memory and `data_end`, a
+pointer to the end of the data in kernel memory. In order to access this data
+and ensure that the eBPF verifier is happy, we'll introduce a helper function
+called `ptr_at`. This function will ensure that before we access any data, we
+check that it's contained between `data` and `data_end`. It is marked as `unsafe`
```

Comment:
ptr_at can actually be made safe: returning a pointer is safe even if the pointer is not aligned, what is unsafe is _deferencing_ the pointer.

Reply:
You've marked this as resolved but in the code ptr_at is still unsafe. It shouldn't be.